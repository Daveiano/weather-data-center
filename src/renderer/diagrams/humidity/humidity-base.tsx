import React, { FunctionComponent, useEffect, useState } from 'react';

import { Loading } from "carbon-components-react";
import { ResponsiveLine } from '@nivo/line'
import { Humidity32 } from "@carbon/icons-react";

import {dataItem, DiagramBaseProps} from "../types";
import { getTimeDifferenceInDays, scaleAverage } from "../scaling";
import { TooltipLine } from "../tooltip";
import { withEmptyCheck } from "../hoc";

const HumidityBase:FunctionComponent<DiagramBaseProps> = (props: DiagramBaseProps): React.ReactElement => {
  const [data, setData] = useState(props.data);
  const [loading, setLoading] = useState(false);
  const [daily, setDaily] = useState(false);

  const scale = () => {
    const timeDifferenceInDays = getTimeDifferenceInDays(props.data);

    let newData: dataItem[];

    // Calculate daily average by summing all measurements an dividing by the
    // number of values.
    if (timeDifferenceInDays > 14) {
      setDaily(true);
      newData = scaleAverage(props.data, 'humidity', 'day');
    } else {
      setDaily(false);
      newData = props.data;
    }

    setData(newData);
    setLoading(false);
  };

  useEffect(() => {
    setLoading(true);
    scale();
  }, [props.data]);

  if (loading) {
    return (
      <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
        <Loading
          description="Active loading indicator"
          withOverlay={false}
        />
      </div>
    );
  }

  return (
    <div data-testid="humidity-diagram">
      <h3>
        <Humidity32 />
        {props.title}
      </h3>

      <div style={{ height: props.height }} className="diagram">
        <ResponsiveLine
          data={[
            {
              id: 'humidity',
              data: data.map(item => ({
                x: item.timeParsed,
                y: item.humidity
              }))
            }
          ]}
          xScale={{
            type: "time",
            useUTC: true,
            format: "%Y-%m-%dT%H:%M:%S.000Z",
            precision: 'minute'
          }}
          xFormat={daily ? "time:%Y/%m/%d" : "time:%Y/%m/%d %H:%M"}
          yScale={{
            type: "linear",
            min: 0,
            max: 103
          }}
          yFormat={value => `${value} %`}
          margin={{ top: 20, right: 10, bottom: 20, left: 40 }}
          curve="cardinal"
          // @todo theme={}
          colors= {['#0099CC']}
          lineWidth={2}
          enableArea={true}
          areaOpacity={0.07}
          enablePoints={true}
          pointSize={5}
          enablePointLabel={false}
          pointLabel="yFormatted"
          axisLeft={{
            legend: '%',
            legendOffset: -35,
            legendPosition: 'middle',
            tickSize: 0,
            tickPadding: 10
          }}
          axisBottom={{
            format: daily ? "%b %Y" : "%e",
            tickValues: daily ? "every month" : "every 3 days",
            tickSize: 0,
            tickPadding: 5
          }}
          isInteractive={true}
          tooltip={point => <TooltipLine point={point.point} />}
          useMesh={true}
          enableCrosshair={true}
        />
      </div>

    </div>
  );
};

export default withEmptyCheck(HumidityBase);