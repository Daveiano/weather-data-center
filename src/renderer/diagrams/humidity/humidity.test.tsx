import React from 'react'
import { render, screen } from '@testing-library/react'

import data from "../../../../tests/data/scaling-input.json";

import { HumidityBase } from "./humidity-base";

test('humidity diagram with 11-day data', async () => {
  const { container } = render(<HumidityBase title="Humidity" height="300px" data={data} />);

  expect(screen.getByTestId('humidity-diagram')).toHaveTextContent("Humidity");

  expect(container).toMatchSnapshot();
});