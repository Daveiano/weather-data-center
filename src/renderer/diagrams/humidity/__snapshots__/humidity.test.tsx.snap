// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`humidity diagram with 11-day data 1`] = `
<div>
  <div
    class="bx--col"
  >
    <div
      class="bx--tile"
    >
      <div
        data-testid="humidity-diagram"
      >
        <h3>
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M23.4761,13.9932,16.8472,3.4365a1.04,1.04,0,0,0-1.6944,0L8.4941,14.0444A9.9861,9.9861,0,0,0,7,19a9,9,0,0,0,18,0A10.0632,10.0632,0,0,0,23.4761,13.9932ZM16,26.0005a7.0089,7.0089,0,0,1-7-7,7.978,7.978,0,0,1,1.2183-3.9438l.935-1.4888L21.2271,23.6411A6.9772,6.9772,0,0,1,16,26.0005Z"
            />
          </svg>
          Humidity
        </h3>
        <div
          class="diagram"
          style="height: 300px;"
        >
          <div
            style="width: 100%; height: 100%;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
